src/server/transaction.cc

```diff
@@ -1060,26 +1077,25 @@ bool Transaction::ScheduleInShard(EngineShard* shard, bool execute_optimistic) {
     // Check if we can run immediately
     if (shard_unlocked && execute_optimistic &&
         CheckLocks(GetDbSlice(shard->shard_id()), mode, lock_args)) {
+      // We need to acquire the fp locks because the executing callback
+      // within RunCallback might preempt.
+      acquire_fp_locks(shard_unlocked);
```

Comment:
I also thought adding a `check` here and make `acquire_fp_locks` return a bool. But there is no really a need since `CheckLocks` gurantees that we hold none.

Reply:
It seems like if we acquire locks, we can remove CheckLocks. Instead the condition will use lock_granted

Reply:
Yes, the code here should be rewritten. We acquire locks in any case, so CheckLocks is not needed